src/charm.py

```diff
@@ -18,6 +22,63 @@ class SpringBootCharm(CharmBase):
     def __init__(self, *args):
         """Initialize the instance."""
         super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self.reconciliation)
+
+    def _spring_boot_container(self) -> ops.model.Container:
+        """Retrieve the container for the Spring Boot application.
+
+        Returns:
+            An instance of :class:`ops.charm.Container` represents the Spring Boot container.
+        """
+        container = self.unit.get_container("spring-boot-app")
+        return container
+
+    def _generate_spring_boot_layer(self) -> dict:
+        """Generate Spring Boot service layer for pebble.
+
+        Returns:
+            Spring Boot service layer configuration, in the form of a dict
+        """
+        container = self._spring_boot_container()
+        if not container.isdir("/app"):
+            raise ReconciliationError(new_status=BlockedStatus("dir /app does not exist"))
+        files_in_app = container.list_files("/app")
+        jar_files = [file.name for file in files_in_app if file.name.endswith(".jar")]
+        if not jar_files:
+            raise ReconciliationError(new_status=BlockedStatus("no jar file exists in /app"))
+        if len(jar_files) > 1:
+            raise ReconciliationError(new_status=BlockedStatus("multiple jar files exist in /app"))
+        jar_file = jar_files[0]
+        return {
```

Comment:
I think the business logic of detecting the type of "jar" and output of the pebble plan should be split in 2 methods like:
1) detect_jar_type
2) create_peblle_plan

Reply:
I separated the jar detection logic into a separate method `_detect_java_application`. Also, create abstractions for different types of java applications in `src/java_application.py`.